shared/src/test/java/org/odk/collect/shared/geometry/GeometryTest.kt

```diff
@@ -0,0 +1,302 @@
+                Point(0.0, 2.0),
+                Point(1.0, 2.0),
+                Point(0.0, 1.0)
+            )
+        )
+
+        assertThat(trace.intersects(), equalTo(true))
+    }
+
+    @Test
+    fun `Trace#intersects returns false for right angled triangle`() {
+        val trace = Trace(
+            listOf(
+                Point(0.0, 0.0),
+                Point(10.0, 10.0),
+                Point(0.0, 10.0)
+            )
+        )
+
+        assertThat(trace.intersects(), equalTo(false))
+    }
+
+    @Test
+    fun `Trace#segments returns false for trace with duplicate points`() {
+        val trace = Trace(
+            listOf(
+                Point(0.0, 0.0),
+                Point(1.0, 0.0),
+                Point(1.0, 0.0),
+                Point(2.0, 0.0)
+            )
+        )
+
+        assertThat(trace.intersects(), equalTo(false))
+    }
+
+    @Test
+    fun `Trace#segments does not include zero-length segments`() {
+        val trace = Trace(
+            listOf(
+                Point(0.0, 0.0),
+                Point(1.0, 0.0),
+                Point(1.0, 0.0),
+                Point(2.0, 0.0)
+            )
+        )
+
+        assertThat(trace.segments(), equalTo(listOf(
+            LineSegment(Point(0.0, 0.0), Point(1.0, 0.0)),
+            LineSegment(Point(1.0, 0.0), Point(2.0, 0.0))
+        )))
+    }
+
+    @Test
+    fun `Trace#intersects returns true for 3 segment closed trace reversing on itself`() {
+        val trace = Trace(listOf(
+            Point(0.0, 0.0),
+            Point(2.0, 0.0),
+            Point(1.0, 0.0),
+            Point(0.0, 0.0)
+        ))
+        assertThat(trace.intersects(), equalTo(true))
+    }
+
+    @Test
+    fun `Trace#intersects satisfies metamorphic relationships`() {
+        repeat(1000) {
+            val trace = generateTrace()
+            val intersects = trace.intersects()
+
+            // Check intersects is consistent when trace is reversed
+            val reversedTrace = Trace(trace.points.reversed())
+            assertThat(
+                "Expected intersects=$intersects:\n$reversedTrace",
+                reversedTrace.intersects(),
+                equalTo(intersects)
+            )
+
+            // Check intersects is consistent when trace is scaled
+            val scaleFactor = Random.nextDouble(0.1, 10.0)
+            val scaledTrace = Trace(trace.points.map {
+                Point(it.x * scaleFactor, it.y * scaleFactor)
+            })
+            assertThat(
+                "Expected intersects=$intersects:\n$scaledTrace",
+                scaledTrace.intersects(),
+                equalTo(intersects)
+            )
+
+            // Check adding an intersection makes intersects true
+            if (!intersects) {
+                val intersectPosition = Random.nextDouble(0.0, 1.0)
+                val intersectionPoint = trace.segments().first().interpolate(intersectPosition)
```

Comment:
Do I understand correctly that if `intersectPosition` is 0, then `trace.segments().first().interpolate(intersectPosition)` returns the first point of the first segment, and as a result, it will form a closed polygon with no intersections?

Reply:
Ah yes! I was meaning to make it use a random segment rather than the first segment (and avoid this). I'll change that.